File: text-justification/text_justification.go
package main

import (
	"strings"
)

func outputOneLine(words []string, width int, lastLine bool) string {
	if len(words) == 1 {
		return words[0] + strings.Repeat(" ", width-len(words[0]))
	}

	if lastLine {
		s := strings.Join(words, " ")
		return s + strings.Repeat(" ", width-len(s))
	}

	var (
		gapCount   = len(words) - 1
		spaceCount int
	)

	spaceCount = width
	for _, word := range words {
		spaceCount -= len(word)
	}

	// 如果需要5个空格, 一共三个 Gap(四个 word), 那么前两个 gap 是两个空格, 后面的一个
	div := spaceCount / gapCount
	mod := spaceCount % gapCount

	rst := make([]string, 0, len(words)+spaceCount)

	for i, word := range words[:len(words)-1] {
		rst = append(rst, word)

		rst = append(rst, strings.Repeat(" ", div))
		if i < mod {
			rst = append(rst, " ")
		}
	}

	rst = append(rst, words[len(words)-1])

	return strings.Join(rst, "")
}

func fullJustify(words []string, maxWidth int) []string {
	rst := make([]string, 0)

	var (
		currentLineWitdh = 0
		currentLineWords = make([]string, 0)
	)
	for _, word := range words {
		if currentLineWitdh+len(word)+len(currentLineWords) <= maxWidth {
			currentLineWords = append(currentLineWords, word)
			currentLineWitdh += len(word)
		} else {
			rst = append(rst, outputOneLine(currentLineWords, maxWidth, false))

			currentLineWords = []string{word}
			currentLineWitdh = len(word)
		}
	}

	if len(currentLineWords) > 0 {
		rst = append(rst, outputOneLine(currentLineWords, maxWidth, true))
	}
	return rst
}

File: text-justification/text-justification.py
"""
https://leetcode.cn/problems/text-justification/
"""

import unittest


class Solution:
    def fillBlank(self, words: list[str], maxWidth: int, last_line: bool) -> str:
        if last_line or len(words) == 1:
            return " ".join(words).ljust(maxWidth)

        blank_count = maxWidth - sum(len(word) for word in words)
        blank_count_for_each = blank_count // (len(words) - 1)
        blanks = [
            (
                1 + blank_count_for_each
                if i < blank_count % (len(words) - 1)
                else blank_count_for_each
            )
            for i in range(len(words) - 1)
        ]
        ans = ""
        for i, word in enumerate(words[:-1]):
            ans += word
            ans += " " * blanks[i]
        return ans + words[-1]

    def fullJustify(self, words: list[str], maxWidth: int) -> list[str]:
        ans = []
        last_word = 0
        length = len(words[0])
        for i in range(1, len(words)):
            word = words[i]
            if length + 1 + len(word) > maxWidth:
                ans.append(self.fillBlank(words[last_word:i], maxWidth, False))
                last_word = i
                length = len(word)
            else:
                length += 1 + len(word)

        if words[last_word:]:
            ans.append(self.fillBlank(words[last_word:], maxWidth, True))
        return ans


class TestSolution(unittest.TestCase):
    def test_fullJustify(self):
        s = Solution()
        rst = s.fullJustify(
            ["This", "is", "an", "example", "of", "text", "justification."], 16
        )
        self.assertEqual(
            rst, ["This    is    an", "example  of text", "justification.  "]
        )

        rst = s.fullJustify(["What", "must", "be", "acknowledgment", "shall", "be"], 16)
        self.assertEqual(
            rst, ["What   must   be", "acknowledgment  ", "shall be        "]
        )
        rst = s.fullJustify(
            [
                "Science",
                "is",
                "what",
                "we",
                "understand",
                "well",
                "enough",
                "to",
                "explain",
                "to",
                "a",
                "computer.",
                "Art",
                "is",
                "everything",
                "else",
                "we",
                "do",
            ],
            20,
        )

        self.assertEqual(
            rst,
            [
                "Science  is  what we",
                "understand      well",
                "enough to explain to",
                "a  computer.  Art is",
                "everything  else  we",
                "do                  ",
            ],
        )


if __name__ == "__main__":
    unittest.main()
